### DISCUSSION TITLE
How can I escape `toJSON` calls?

### DISCUSSION THREAD

**Discussion by jhpratt** [community/community#34563] [Actions] [answered]
I have a part of a matrix that I would like to stringify as JSON for some processing. However, this processing will be done with bash. I need a way to escape the double quotes in the returned value from `toJSON`. Is this even possible? I would use single quotes for the bash string, but this value could potentially have single quotes in it (necessitating escaping of its own).

TL;DR turn `["foo"]` into `[\"foo\"]`, such that I can put the resulting value in a bash string.

**Comment by airtower-luna** [comment, marked as answer]
That sounds like you're embedding the `toJSON()` call directly in the script, like so:

```yaml
run: echo "${{ toJSON(matrix.something) }}"
```

The answer is: Don't do that. The Actions syntax `${{ ... }}` is substituted *literally* before Bash ever sees the script, so Bash has no way of knowing what isn't meant to be syntax.

Instead, use an environment variable:

```yaml
run: echo "${JSON}"
env:
  JSON: ${{ toJSON(matrix.something) }}
```

That way Bash is aware of the variable and can handle it accordingly.

**Reply by jhpratt** [reply]
> The answer is: Don't do that. The Actions syntax ${{ ... }} is substituted literally before Bash ever sees the script, so Bash has no way of knowing what isn't meant to be syntax.

That's precisely what I thought, hence my confusion.

> Instead, use an environment variable:

I had previously tried this locally and it failed, presumably due to string escaping issues of its own. It works on CI, though, so :shrug: 

Thank you!
